pkg/parse/updater.go

```diff
@@ -214,6 +202,20 @@ func (u *Updater) update(ctx context.Context, cache *cacheForCommit) status.Mult
 		}
 	}
 
+	// Apply the declared resources
+	if !cache.applied {
+		declaredObjs, _ := u.Resources.DeclaredObjects()
+		_, err := u.apply(ctx, declaredObjs, cache.source.commit)
```

Comment:
The watches are by GVK, so we can't necessarily stop watches for individual objects that fail to be applied (unless _all_ objects of a given GVK failed to apply). We could potentially ignore watch events for a object which failed to apply, although I'm not sure what the value add would be for that new behavior.

What is the motivation for the behavior you propose?

Reply:
I am concerned that we may miss some events.

- For actuation, it works alright. If an object failed to be applied, while the watch is still running. Any drift will be corrected by the remediator.
- For pruning, when the apply (prune) failed, the watch was not stopped before, but it is stopped with the change. Any drift before the next retry won't get remediated.

Reply:
This problem isn't solvable with the current design. If we update the watches before the apply, they will fail on new resource types and stop watching pruned resources prematurely. If we update the watched after the apply, we might miss events. When I refactored to add remediator pausing during apply, I opted to update the watched after the apply because missing events was the lesser evil.

The only options I know of to mitigate this are:
1. duplicate the watch code in the remediator to act like the applier watches, which watch CRDs and add/remove watches dynamically (more resource usage and complexity)
2. merge the applier and remediator into a single continuous reconciler (b/205185055)

Reply:
> For pruning, when the apply (prune) failed, the watch was not stopped before, but it is stopped with the change. Any drift before the next retry won't get remediated.

> If we update the watches before the apply, we miss events for pruned resource types and error on new resource types.

I don't think these statements are strictly true. We pause the remediator and update the declared resource cache before applying, so we will miss events for pruned resources types regardless of when the watches start. We would need to change the logic for the declared resource cache as well if we want to continue remediating pruned objects after a failed apply.

The updated implementation will handle new resource types in the following way: on the first pass of adding a new resource type the watches will fail and the resources will still be applied. Because there was an error setting up the watches, the reconciler will retry. On the second pass, the watches should succeed to be created and the applier will re-run - at which point the remediator can correct for drift with a clean initial state (because the watches were started before the applier ran the second time).